File: app/src/main/java/com/simplertutorials/android/wheathograophy/data/database/DatabaseRepository.java
package com.simplertutorials.android.wheathograophy.data.database;

import android.annotation.SuppressLint;
import android.content.SharedPreferences;
import com.simplertutorials.android.wheathograophy.domain.City;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

public class DatabaseRepository {
    private static DatabaseRepository instance;
    private SharedPreferences settings;
    private SharedPreferences.Editor editor;
    private DatabaseManager manager;
    private String KEY;

    @SuppressLint("CommitPrefEdits")
    private DatabaseRepository(SharedPreferences settings, String key) {
        this.settings = settings;
        this.editor = settings.edit();
        this.manager = new DatabaseManager(editor,settings);
        this.KEY = key;
    }

    public static DatabaseRepository getInstance(SharedPreferences settings, String key) {
        if (instance == null)
            instance = new DatabaseRepository(settings, key);
        return instance;
    }

    public ArrayList<City>getCityList(){
        return convertToCityObject(manager.readSet(KEY));
    }

    public void addCity(City city) {
        ArrayList cityList = getCityList();
        if (!(cityList.contains(city))) {
            cityList.add(city);
            rewriteSharedPreferences(cityList);
        }
    }

    public void deleteCity(City city) {
        ArrayList cityList = getCityList();
        if (cityList.contains(city)) {
            cityList.remove(city);
            rewriteSharedPreferences(cityList);
        }
    }

    private void rewriteSharedPreferences(ArrayList<City> cityList) {
        manager.deleteValue(KEY);
        manager.writeSet(convertToHashSet(cityList), KEY);
    }

    private HashSet<String> convertToHashSet(ArrayList<City> cityList){
        HashSet<String> hashSet = new HashSet<String>();
        for (City it: cityList) {
            hashSet.add(it.getName());
        }
        return hashSet;
    }

    private ArrayList<City> convertToCityObject(Set<String> readSet){
        ArrayList cityList = new ArrayList<City>();

        if (readSet != null) {
            for (String it : readSet){
                City city = new City(it);
                cityList.add(city);
            }
        }
        return cityList;
    }

}

File: app/src/main/java/com/simplertutorials/android/wheathograophy/data/database/DatabaseRepository.kt
package com.simplertutorials.android.wheathograophy.data.database

import android.annotation.SuppressLint
import android.content.SharedPreferences
import com.simplertutorials.android.wheathograophy.domain.City

class DatabaseRepository @SuppressLint("CommitPrefEdits") private constructor(
    private val settings: SharedPreferences,
    key: String
) {
    private val editor: SharedPreferences.Editor = settings.edit()
    private val manager: DatabaseManager = DatabaseManager(editor, settings)
    private val KEY: String = key
    val cityList: List<City>
        get() = convertToCityObject(manager.readSet(KEY))

    fun addCity(city: City) {
        val cityList = mutableListOf<City>()
        cityList.addAll(cityList)
        if (!cityList.contains(city)) {
            cityList.add(city)
            rewriteSharedPreferences(cityList)
        }
    }

    fun deleteCity(city: City?) {
        val cityList = mutableListOf<City>()
        cityList.addAll(cityList)
        if (cityList.contains(city)) {
            cityList.remove(city)
            rewriteSharedPreferences(cityList)
        }
    }

    private fun rewriteSharedPreferences(cityList: List<City>) {
        manager.deleteValue(KEY)
        manager.writeSet(convertToHashSet(cityList), KEY)
    }

    private fun convertToHashSet(cityList: List<City>): HashSet<String> {
        val hashSet = HashSet<String>()
        for (it in cityList) {
            hashSet.add(it.name)
        }
        return hashSet
    }

    private fun convertToCityObject(readSet: Set<String>?): List<City> {
        val cityList = mutableListOf<City>()
        if (readSet != null) {
            for (it in readSet) {
                val city = City(it)
                cityList.add(city)
            }
        }
        return cityList
    }

    companion object {
        private var instance: DatabaseRepository? = null
        fun getInstance(settings: SharedPreferences, key: String): DatabaseRepository {
            if (instance == null) instance = DatabaseRepository(settings, key)
            return instance!!
        }
    }

}
